File: src/heimdall/contract.py
from __future__ import annotations

import hashlib
import json
import re
import sys
from dataclasses import dataclass
from importlib import resources
from pathlib import Path
from typing import Any

import yaml
from jsonschema import Draft202012Validator


class ContractError(ValueError):
    def __init__(self, code: str):
        self.code = code
        super().__init__(code)


class UniqueKeyLoader(yaml.SafeLoader):
    pass


def _construct_unique_mapping(loader: UniqueKeyLoader, node: yaml.MappingNode, deep: bool = False):
    mapping: dict[Any, Any] = {}
    for key_node, value_node in node.value:
        key = loader.construct_object(key_node, deep=deep)
        if key in mapping:
            raise ContractError("duplicate_key")
        mapping[key] = loader.construct_object(value_node, deep=deep)
    return mapping


UniqueKeyLoader.add_constructor(
    yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG, _construct_unique_mapping
)


@dataclass(frozen=True)
class EvalSpec:
    manifest_path: Path
    target_id: str
    target_root: Path
    policy: dict[str, Any]
    policy_path: Path
    policy_digest: str
    isolation: str
    argv: tuple[str, ...]
    cwd: str
    timeout_seconds: int
    env: dict[str, str]
    checks: tuple[dict[str, Any], ...]


def _load_yaml(path: Path) -> Any:
    try:
        return yaml.load(path.read_text(encoding="utf-8"), Loader=UniqueKeyLoader)
    except ContractError:
        raise
    except (OSError, UnicodeError, yaml.YAMLError) as error:
        raise ContractError("invalid_manifest") from error


def _schema_bytes(name: str) -> bytes:
    candidates = [
        Path(__file__).resolve().parents[2] / "schemas" / name,
        Path(sys.prefix) / "share" / "heimdall" / "schemas" / name,
    ]
    for candidate in candidates:
        if candidate.is_file():
            return candidate.read_bytes()
    try:
        return resources.files("heimdall").joinpath("schemas", name).read_bytes()
    except (FileNotFoundError, ModuleNotFoundError) as error:
        raise ContractError("schema_unavailable") from error


def _validate_schema(document: Any, schema_name: str) -> None:
    try:
        schema = json.loads(_schema_bytes(schema_name))
        Draft202012Validator(schema).validate(document)
    except ContractError:
        raise
    except Exception as error:
        raise ContractError("invalid_manifest") from error


def _safe_relative(value: str) -> str:
    path = Path(value)
    if path.is_absolute() or ".." in path.parts:
        raise ContractError("invalid_manifest")
    return value


def _load_policy(path: Path) -> dict[str, Any]:
    policy = _load_yaml(path)
    if not isinstance(policy, dict) or set(policy) != {
        "schema_version", "id", "version", "criteria"
    }:
        raise ContractError("invalid_policy")
    if policy["schema_version"] != "1.0" or not isinstance(policy["criteria"], list):
        raise ContractError("invalid_policy")
    expected_ids = {
        "contract_fidelity", "authority_safety", "outcome_evidence", "failure_honesty"
    }
    observed: set[str] = set()
    for criterion in policy["criteria"]:
        if (
            not isinstance(criterion, dict)
            or set(criterion) != {"id", "required"}
            or criterion.get("id") not in expected_ids
            or not isinstance(criterion.get("required"), bool)
            or criterion["id"] in observed
        ):
            raise ContractError("invalid_policy")
        observed.add(criterion["id"])
    if observed != expected_ids:
        raise ContractError("invalid_policy")
    return policy


def load_spec(path: str | Path) -> EvalSpec:
    manifest_path = Path(path).expanduser().resolve()
    document = _load_yaml(manifest_path)
    _validate_schema(document, "eval-spec.v1.json")

    target_root_value = _safe_relative(document["target"]["root"])
    cwd = _safe_relative(document["command"].get("cwd", "."))
    check_ids: set[str] = set()
    checks: list[dict[str, Any]] = []
    for raw in document["checks"]:
        check = dict(raw)
        _safe_relative(check["path"])
        if check["id"] in check_ids:
            raise ContractError("invalid_manifest")
        check_ids.add(check["id"])
        if check["kind"] == "file_equals" and "expected" not in check:
            raise ContractError("invalid_manifest")
        if check["kind"] != "file_equals" and "expected" in check:
            raise ContractError("invalid_manifest")
        checks.append(check)

    target_root = (manifest_path.parent / target_root_value).resolve()
    if not target_root.is_dir():
        raise ContractError("target_unavailable")

    policy_path_value = document["policy"]["path"]
    if Path(policy_path_value).is_absolute():
        raise ContractError("invalid_policy")
    policy_path = (manifest_path.parent / policy_path_value).resolve()
    if not policy_path.is_file():
        raise ContractError("invalid_policy")
    policy = _load_policy(policy_path)
    if (
        policy["id"] != document["policy"]["id"]
        or policy["version"] != document["policy"]["version"]
    ):
        raise ContractError("policy_mismatch")

    env = dict(document["command"].get("env", {}))
    for key in env:
        if re.fullmatch(r"[A-Z_][A-Z0-9_]{0,63}", key) is None:
            raise ContractError("invalid_manifest")

    return EvalSpec(
        manifest_path=manifest_path,
        target_id=document["target"]["id"],
        target_root=target_root,
        policy=policy,
        policy_path=policy_path,
        policy_digest=hashlib.sha256(policy_path.read_bytes()).hexdigest(),
        isolation=document["isolation"],
        argv=tuple(document["command"]["argv"]),
        cwd=cwd,
        timeout_seconds=document["command"]["timeout_seconds"],
        env=env,
        checks=tuple(checks),
    )

File: internal/contract/contract.go
package contract

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/JeremyDev87/heimdall/internal/schema"
	"go.yaml.in/yaml/v3"
)

type Error struct{ code string }

func (err *Error) Error() string { return err.code }
func fail(code string) error     { return &Error{code: code} }
func Code(err error) string {
	var contractError *Error
	if errors.As(err, &contractError) {
		return contractError.code
	}
	return ""
}

type Check struct {
	ID       string
	Kind     string
	Path     string
	Expected *string
}

type Spec struct {
	ManifestPath   string
	TargetID       string
	TargetRoot     string
	Policy         map[string]any
	PolicyPath     string
	PolicyDigest   string
	Isolation      string
	Argv           []string
	Cwd            string
	TimeoutSeconds int
	Env            map[string]string
	Checks         []Check
}

type wireSpec struct {
	SchemaVersion string `json:"schema_version"`
	Target        struct {
		ID   string `json:"id"`
		Root string `json:"root"`
	} `json:"target"`
	Policy struct {
		ID      string `json:"id"`
		Version string `json:"version"`
		Path    string `json:"path"`
	} `json:"policy"`
	Isolation string `json:"isolation"`
	Command   struct {
		Argv           []string          `json:"argv"`
		Cwd            string            `json:"cwd"`
		TimeoutSeconds int               `json:"timeout_seconds"`
		Env            map[string]string `json:"env"`
	} `json:"command"`
	Checks []struct {
		ID       string  `json:"id"`
		Kind     string  `json:"kind"`
		Path     string  `json:"path"`
		Expected *string `json:"expected"`
	} `json:"checks"`
}

var envName = regexp.MustCompile(`^[A-Z_][A-Z0-9_]{0,63}$`)

func LoadSpec(path string) (*Spec, error) {
	manifestPath, err := filepath.Abs(expandHome(path))
	if err != nil {
		return nil, fail("invalid_manifest")
	}
	manifestPath, err = filepath.EvalSymlinks(manifestPath)
	if err != nil {
		return nil, fail("invalid_manifest")
	}
	document, _, err := loadYAML(manifestPath)
	if err != nil {
		return nil, err
	}
	if err := schema.Validate("eval-spec.v1.json", document); err != nil {
		return nil, fail("invalid_manifest")
	}
	data, err := json.Marshal(document)
	if err != nil {
		return nil, fail("invalid_manifest")
	}
	var wire wireSpec
	if err := json.Unmarshal(data, &wire); err != nil {
		return nil, fail("invalid_manifest")
	}

	if err := safeRelative(wire.Target.Root); err != nil {
		return nil, err
	}
	cwd := wire.Command.Cwd
	if cwd == "" {
		cwd = "."
	}
	if err := safeRelative(cwd); err != nil {
		return nil, err
	}
	checkIDs := map[string]bool{}
	checks := make([]Check, 0, len(wire.Checks))
	for _, raw := range wire.Checks {
		if err := safeRelative(raw.Path); err != nil {
			return nil, err
		}
		if checkIDs[raw.ID] {
			return nil, fail("invalid_manifest")
		}
		checkIDs[raw.ID] = true
		if raw.Kind == "file_equals" && raw.Expected == nil {
			return nil, fail("invalid_manifest")
		}
		if raw.Kind != "file_equals" && raw.Expected != nil {
			return nil, fail("invalid_manifest")
		}
		checks = append(checks, Check{ID: raw.ID, Kind: raw.Kind, Path: raw.Path, Expected: raw.Expected})
	}

	targetRoot := filepath.Clean(filepath.Join(filepath.Dir(manifestPath), wire.Target.Root))
	targetRoot, err = filepath.EvalSymlinks(targetRoot)
	if err != nil {
		return nil, fail("target_unavailable")
	}
	info, err := os.Stat(targetRoot)
	if err != nil || !info.IsDir() {
		return nil, fail("target_unavailable")
	}
	if filepath.IsAbs(wire.Policy.Path) {
		return nil, fail("invalid_policy")
	}
	policyPath := filepath.Clean(filepath.Join(filepath.Dir(manifestPath), wire.Policy.Path))
	policyPath, err = filepath.EvalSymlinks(policyPath)
	if err != nil {
		return nil, fail("invalid_policy")
	}
	info, err = os.Stat(policyPath)
	if err != nil || !info.Mode().IsRegular() {
		return nil, fail("invalid_policy")
	}
	policy, policyBytes, err := loadPolicy(policyPath)
	if err != nil {
		return nil, err
	}
	if policy["id"] != wire.Policy.ID || policy["version"] != wire.Policy.Version {
		return nil, fail("policy_mismatch")
	}
	for key := range wire.Command.Env {
		if !envName.MatchString(key) {
			return nil, fail("invalid_manifest")
		}
	}
	if wire.Command.Env == nil {
		wire.Command.Env = map[string]string{}
	}
	sum := sha256.Sum256(policyBytes)
	return &Spec{
		ManifestPath: manifestPath, TargetID: wire.Target.ID, TargetRoot: targetRoot,
		Policy: policy, PolicyPath: policyPath, PolicyDigest: hex.EncodeToString(sum[:]),
		Isolation: wire.Isolation, Argv: wire.Command.Argv, Cwd: cwd,
		TimeoutSeconds: wire.Command.TimeoutSeconds, Env: wire.Command.Env, Checks: checks,
	}, nil
}

func loadPolicy(path string) (map[string]any, []byte, error) {
	document, data, err := loadYAML(path)
	if err != nil {
		return nil, nil, err
	}
	policy, ok := document.(map[string]any)
	if !ok || !exactKeys(policy, "schema_version", "id", "version", "criteria") {
		return nil, nil, fail("invalid_policy")
	}
	criteria, ok := policy["criteria"].([]any)
	if policy["schema_version"] != "1.0" || !ok {
		return nil, nil, fail("invalid_policy")
	}
	expected := map[string]bool{"contract_fidelity": true, "authority_safety": true, "outcome_evidence": true, "failure_honesty": true}
	observed := map[string]bool{}
	for _, raw := range criteria {
		criterion, ok := raw.(map[string]any)
		if !ok || !exactKeys(criterion, "id", "required") {
			return nil, nil, fail("invalid_policy")
		}
		id, idOK := criterion["id"].(string)
		_, requiredOK := criterion["required"].(bool)
		if !idOK || !requiredOK || !expected[id] || observed[id] {
			return nil, nil, fail("invalid_policy")
		}
		observed[id] = true
	}
	if len(observed) != len(expected) {
		return nil, nil, fail("invalid_policy")
	}
	return policy, data, nil
}

func loadYAML(path string) (any, []byte, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, nil, fail("invalid_manifest")
	}
	var node yaml.Node
	decoder := yaml.NewDecoder(bytes.NewReader(data))
	if err := decoder.Decode(&node); err != nil {
		return nil, nil, fail("invalid_manifest")
	}
	if len(node.Content) == 0 {
		return nil, nil, fail("invalid_manifest")
	}
	if duplicateKeys(node.Content[0]) {
		return nil, nil, fail("duplicate_key")
	}
	var raw any
	if err := node.Content[0].Decode(&raw); err != nil {
		return nil, nil, fail("invalid_manifest")
	}
	normalized, err := normalizeYAML(raw)
	if err != nil {
		return nil, nil, fail("invalid_manifest")
	}
	jsonBytes, err := json.Marshal(normalized)
	if err != nil {
		return nil, nil, fail("invalid_manifest")
	}
	var document any
	jsonDecoder := json.NewDecoder(bytes.NewReader(jsonBytes))
	jsonDecoder.UseNumber()
	if err := jsonDecoder.Decode(&document); err != nil {
		return nil, nil, fail("invalid_manifest")
	}
	return document, data, nil
}

func duplicateKeys(node *yaml.Node) bool {
	if node.Kind == yaml.MappingNode {
		seen := map[string]bool{}
		for index := 0; index < len(node.Content); index += 2 {
			key := node.Content[index]
			if key.Kind != yaml.ScalarNode || seen[key.Value] {
				return true
			}
			seen[key.Value] = true
			if duplicateKeys(node.Content[index+1]) {
				return true
			}
		}
		return false
	}
	for _, child := range node.Content {
		if duplicateKeys(child) {
			return true
		}
	}
	return false
}

func normalizeYAML(value any) (any, error) {
	switch typed := value.(type) {
	case map[string]any:
		result := make(map[string]any, len(typed))
		for key, item := range typed {
			normalized, err := normalizeYAML(item)
			if err != nil {
				return nil, err
			}
			result[key] = normalized
		}
		return result, nil
	case map[any]any:
		result := make(map[string]any, len(typed))
		for key, item := range typed {
			text, ok := key.(string)
			if !ok {
				return nil, fail("invalid_manifest")
			}
			normalized, err := normalizeYAML(item)
			if err != nil {
				return nil, err
			}
			result[text] = normalized
		}
		return result, nil
	case []any:
		result := make([]any, len(typed))
		for index, item := range typed {
			normalized, err := normalizeYAML(item)
			if err != nil {
				return nil, err
			}
			result[index] = normalized
		}
		return result, nil
	default:
		return value, nil
	}
}

func safeRelative(value string) error {
	if filepath.IsAbs(value) || filepath.VolumeName(value) != "" {
		return fail("invalid_manifest")
	}
	for _, part := range strings.FieldsFunc(filepath.ToSlash(value), func(r rune) bool { return r == '/' }) {
		if part == ".." {
			return fail("invalid_manifest")
		}
	}
	return nil
}

func exactKeys(value map[string]any, expected ...string) bool {
	if len(value) != len(expected) {
		return false
	}
	sort.Strings(expected)
	actual := make([]string, 0, len(value))
	for key := range value {
		actual = append(actual, key)
	}
	sort.Strings(actual)
	for index := range actual {
		if actual[index] != expected[index] {
			return false
		}
	}
	return true
}

func expandHome(path string) string {
	if path == "~" || strings.HasPrefix(path, "~/") {
		if home, err := os.UserHomeDir(); err == nil {
			return filepath.Join(home, strings.TrimPrefix(path, "~/"))
		}
	}
	return path
}
